pkg/healthv2/provider.go

```diff
@@ -0,0 +1,219 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package healthv2
+
+import (
+	"context"
+	"fmt"
+	"sync/atomic"
+	"time"
+
+	"github.com/cilium/cilium/pkg/healthv2/types"
+	"github.com/cilium/cilium/pkg/hive/cell"
+	"github.com/cilium/cilium/pkg/statedb"
+	"github.com/cilium/cilium/pkg/statedb/index"
+	"github.com/sirupsen/logrus"
+)
+
+type HealthProviderParams struct {
+	cell.In
+	Lifecycle cell.Lifecycle
+	DB        *statedb.DB
+}
+
+type HealthProvider struct {
+	db           *statedb.DB
+	stopped      atomic.Bool
+	statusTable  statedb.RWTable[types.Status]
+	Statuses     statedb.Table[types.Status]
+	PrimaryIndex statedb.Index[types.Status, types.HealthID]
+	LevelIndex   statedb.Index[types.Status, types.Level]
+}
+
+const HealthTable = "health"
+
+func newHealthV2Provider(params HealthProviderParams) (types.Provider, *HealthProvider, error) {
+	p := &HealthProvider{
+		db: params.DB,
+	}
+	p.PrimaryIndex = statedb.Index[types.Status, types.HealthID]{
+		Name: "identifier",
+		FromObject: func(s types.Status) index.KeySet {
+			return index.NewKeySet([]byte(s.ID.String()))
+		},
+		FromKey: func(k types.HealthID) index.Key {
+			return index.Key([]byte(k))
+		},
+		Unique: true,
+	}
+	p.LevelIndex = statedb.Index[types.Status, types.Level]{
+		Name: "level",
+		FromObject: func(s types.Status) index.KeySet {
+			return index.NewKeySet([]byte(s.Level))
+		},
+		FromKey: func(key types.Level) index.Key {
+			return index.Key([]byte(key))
+		},
+		Unique: false,
+	}
+	var err error
+	p.statusTable, err = statedb.NewTable(HealthTable,
+		p.PrimaryIndex,
+		p.LevelIndex)
+	p.Statuses = p.statusTable
+	if err != nil {
+		return nil, nil, err
+	}
+	if err := p.db.RegisterTable(p.statusTable); err != nil {
+		return nil, nil, err
+	}
+
+	params.Lifecycle.Append(p)
+
+	return p, p, nil
+}
+
+func (p *HealthProvider) Start(ctx cell.HookContext) error {
+	return nil
+}
+
+func (p *HealthProvider) Stop(ctx cell.HookContext) error {
+	p.stopped.Store(true)
+	tx := p.db.ReadTxn()
+	iter, _ := p.statusTable.All(tx)
+	for {
+		s, rev, ok := iter.Next()
+		if !ok {
+			break
+		}
+		logrus.Infof("%s (rev=%d)", s.ID.String(), rev)
```

Comment:
Shouldn't we want to use a FieldLogger provided by hive? Or pkg/logging ?

Reply:
Yup, good catch, this was supposed to be a temporary measure while developing the code, ill switch that over.